transformer_engine/pytorch/attention.py

```diff
@@ -1841,24 +2062,49 @@ def forward(
                 and (core_attention_bias_type == "no_bias")
                 and (fused_attention_backend
                     == tex.NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen))
-        with self.attention_dropout_ctx():
-            output = FusedAttnFunc.apply(
-                self.training,
-                max_seqlen_q, max_seqlen_kv,
-                cu_seqlens_q, cu_seqlens_kv,
-                query_layer, key_layer, value_layer,
-                qkv_dtype,
-                core_attention_bias,
-                1.0/self.norm_factor,
-                self.attention_dropout if self.training else 0.0,
-                fast_zero_fill,
-                qkv_layout,
-                core_attention_bias_type,
-                attn_mask_type,
-                None, # rng_gen
-                fused_attention_backend,
-                use_FAv2_bwd,
-            )
+
+        if context_parallel:
+            assert (fused_attention_backend
+                == tex.NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen
+                ), "Context parallelism is only supported with backend of NVTE_F16_arbitrary_seqlen!"
```

Comment:
We should add a check in `DotProductAttention` to screen the backend returned for `FusedAttention`. If the backend is not `tex.NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen`, we shouldn't let the code run into the `FusedAttention` module. Something like this
```
context_parallel = (cp_group is not None) and (get_distributed_world_size(cp_group) != 1)
if context_parallel and fused_attention_backend != tex.NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen:
    use_fused_attention = False
```
here: https://github.com/NVIDIA/TransformerEngine/blob/1bb8b6ebae1ecadd7d817d572245ce5243166d32/transformer_engine/pytorch/attention.py#L2385

Also, does CP really not work with the `NVTE_F16_max512` backend?